core/src/main/scala/quasar/plugin/postgres/CsvUpsertSink.scala

```diff
@@ -0,0 +1,196 @@
+        case Some(t) =>
+          t.pure[F]
+
+        case None =>
+          MonadResourceErr[F].raiseError(ResourceError.notAResource(args.path))
+      }
+
+
+    def handleCreate(totalBytes: Ref[F, Long], records: Chunk[Byte]): ConnectionIO[Unit] =
+      for {
+        tbl <- toConnectionIO(table)
+
+        colSpecs <- toConnectionIO(specifyColumnFragments[F](columns))
+
+        indexColumn = Fragments.parentheses(Fragment.const(hygienicIdent(args.idColumn.name)))
+
+        _ <- (args.writeMode, writeMode) match {
+          case (QWriteMode.Replace, WriteMode.Create) =>
+            createTable(log)(tbl, colSpecs) >> createIndex(log)(tbl, indexColumn)
+          case (QWriteMode.Replace, WriteMode.Replace) =>
+            dropTableIfExists(log)(tbl) >> createTable(log)(tbl, colSpecs) >> createIndex(log)(tbl, indexColumn)
+          case (QWriteMode.Replace, WriteMode.Truncate) =>
+            createTableIfNotExists(log)(tbl, colSpecs) >> truncateTable(log)(tbl) >> createIndex(log)(tbl, indexColumn)
+          case (QWriteMode.Replace, WriteMode.Append) =>
+            createTableIfNotExists(log)(tbl, colSpecs) >> createIndex(log)(tbl, indexColumn)
+          case (QWriteMode.Append, _) =>
+            ().pure[ConnectionIO]
+        }
+
+        cols = columns.map(c => hygienicIdent(c.name)).intercalate(", ")
+
+        copyQuery =
+          s"COPY ${hygienicIdent(tbl)} ($cols) FROM STDIN WITH (FORMAT csv, HEADER FALSE, ENCODING 'UTF8')"
+
+        data = records.toBytes
+
+        copied = PFCM.copyIn(copyQuery).bracketCase(
+          PFCM.embed(_, PFCI.writeToCopy(data.values, data.offset, data.length))) { (pgci, exitCase) =>
+          PFCM.embed(pgci, exitCase match {
+            case ExitCase.Completed => PFCI.endCopy.void
+            case _ => PFCI.isActive.ifM(PFCI.cancelCopy, PFCI.unit)
+          })
+        }
+
+        back <- PHC.pgGetCopyAPI(copied)
+
+        _ <- toConnectionIO(recordChunks[F](totalBytes, log)(records))
+
+      } yield back
+
+    def handleDelete(recordIds: IdBatch): ConnectionIO[Unit] =
+      if (recordIds.size === 0) {
+        ().pure[ConnectionIO]
+      } else {
+
+        def preambleFragment(tbl: Table): Fragment =
+          fr"DELETE FROM" ++
+            Fragment.const(hygienicIdent(tbl)) ++
+            fr"WHERE" ++
+            Fragment.const(hygienicIdent(args.idColumn.name))
+
+        def deleteFrom(preamble: Fragment, table: Table): ConnectionIO[Int] =
+          recordIds match {
+            case IdBatch.Strings(values, _) =>
+              Fragments.in(preamble, NonEmptyVector.fromVectorUnsafe(values.toVector)) // trust size passed by quasar
+                .updateWithLogHandler(logHandler(log))
+                .run
+            case IdBatch.Longs(values, _) =>
+              Fragments.in(preamble, NonEmptyVector.fromVectorUnsafe(values.toVector))
+                .updateWithLogHandler(logHandler(log))
+                .run
+            case IdBatch.Doubles(values, _) =>
```

Comment:
This part of the API is a bit strange. Does anyone use floats/doubles as primary keys? Seems ill-advised.

Reply:
@jsantos17 what are the concerns you're thinking of, regarding why this would be ill-advised?

Reply:
Because of rounding it might be possible to have several representations of the same number. But anyway, after researching a bit I find people do use floats for primary keys so we should support it.

Reply:
👍 hopefully if float is being used as a primary key, they will already be represented consistently in the underlying datastore, meaning that when we parse it out, we'll get a consistent representation of each value.